system/lib/wasmfs/backends/opfs_backend.cpp

```diff
@@ -118,6 +118,32 @@ void _wasmfs_opfs_flush_access(em_proxying_ctx* ctx, int access_id, int* err);
 namespace {
 
 using ProxyWorker = emscripten::ProxyWorker;
+using ProxyingQueue = emscripten::ProxyingQueue;
+
+class Worker {
+public:
+#ifdef __EMSCRIPTEN_PTHREADS__
+  ProxyWorker proxy;
+
+  void operator()(const std::function<void()>& func) {
+    proxy(func);
+  }
+  void operator()(const std::function<void(ProxyingQueue::ProxyingCtx)>& func) {
+    proxy(func);
+  }
+#else
+  // When used with JSPI on the main thread the various wasmfs_opfs_* functions
+  // can be directly executed since they are all async.
+  void operator()(const std::function<void()>& func) {
+    func();
+  }
```

Comment:
I think we should be able to make these templates to avoid having to actually wrap the lambda in a `std::function`. Since we need to handle different signatures differently, you might also need to use `std::is_invokable` to selectively enable the templates.

This might not be worth the effort.

Reply:
Looks like `std::is_invocable` is only on c++17, which I think we're still targeting 14. I don't really see a clean way of doing this without that.

Reply:
WasmFS is built with C++17: https://github.com/emscripten-core/emscripten/blob/main/tools/system_libs.py#L1764